modules/internal/String.chpl

```diff
@@ -17,572 +17,1436 @@
-    return x:string + s;
+      if this.locale.id != chpl_nodeID then
+        halt("Cannot call .c_str() on a remote string");
 
-  proc typeToString(type t) param {
-    return __primitive("typeToString", t);
-  }
-  
-  proc typeToString(x) param {
-    compilerError("typeToString()'s argument must be a type, not a value");
-  }
+      if debugStrings then
+        chpl_debug_string_print("leaving .c_str()");
 
-  // This implies that the *representation* of strings is shared.
-  // If strings are reimplemented as classes or records, a less trivial
-  // implementation for assignment will become necessary.
-  inline proc =(ref a: string, b: string) { __primitive("move", a, b); }
+      return this.buff:c_string;
+    }
 
-  inline proc ==(a: string, b: string) return (__primitive("string_compare", a.c_str(), b.c_str()) == 0);
-  inline proc !=(a: string, b: string) return (__primitive("string_compare", a.c_str(), b.c_str()) != 0);
+    inline proc length return len;
+    //inline proc size return len;
+
+    // Localize gets a version of this that is on the currently executing
+    // locale. If this is already on the current locale we return a shallow
+    // copy, otherwise a remote copy is performed.
+    inline proc localize() /*: string*/ {
+        const ret: string = if this.locale.id != chpl_nodeID
+          then this // assignment makes it local
+          else new string(this, owned=false);
+        return ret;
+    }
 
 
-  inline proc <=(a: string, b: string) return a.c_str()<=b.c_str();
-  inline proc >=(a: string, b: string) return a.c_str()>=b.c_str();
-  inline proc <(a: string, b: string) return a.c_str()<b.c_str();
-  inline proc >(a: string, b: string) return a.c_str()>b.c_str();
+    // Steals ownership of string.buff. This is unsafe and you probably don't
+    // want to use it. I'd like to figure out how to get rid of it entirely.
+    proc _steal_buffer() /*: bufferType*/ {
+      const ret = this.buff;
+      this.buff = nil;
+      this.len = 0;
+      this._size = 0;
+      return ret;
+    }
 
-  //
-  // primitive string functions and methods
-  //
-  inline proc ascii(a: string) return ascii(a.c_str());
-  inline proc string.length return this.c_str().length;
-  inline proc string.size return this.length;
-  inline proc string.substring(i: int) {
-    var cs = this.c_str().substring(i);
-    const ret = toString(cs);
-    // toString steals the returned c_string_copy, so no need to free it.
-    return ret;
-  }
-  inline proc string.substring(r: range(?)) {
-    var cs = this.c_str().substring(r);
-    const ret = toString(cs);
-    // toString steals the returned c_string_copy, so no need to free it.
+    // TODO: cant explicitly state return type right now due to a bug in the
+    // compiler. Uncomment this funciton and others when possible.
+    iter these() /*: string*/ {
+      for i in 1..this.len {
+        yield this[i];
+      }
+    }
+
+    proc this(i: int) /*: string*/ {
```

Comment:
What is your intention regarding Unicode and this routine? Could we have a comment that says what it does now and what it will do (if we know and if it's different) once Unicode is better supported?